nx/lib/nx/shape.ex

```diff
@@ -1218,6 +1218,57 @@ defmodule Nx.Shape do
     {List.to_tuple(result_shape), result_names}
   end
 
+  @doc """
+  Returns shape if valid and raises error if not.
+  """
+  def take_diagonal(tensor) do
+    case tensor.shape do
+      {len, breadth} ->
+        {len, breadth}
+
+      _bad_shape ->
+        raise ArgumentError, "take_diagonal/2 expects tensor of rank 2, got:\n#{inspect(tensor)}"
+    end
+  end
+
+  @doc """
+  Validates an offset for given shape
+
+  ## Examples
+
+      iex> Nx.Shape.validate_offset!({3, 4}, 1)
+      :ok
+
+      iex> Nx.Shape.validate_offset!({3, 4}, -1)
+      :ok
+
+  ## Error cases
+
+    iex> Nx.Shape.validate_offset!({3, 4}, 3)
+    ** (ArgumentError) offset must be less than length when positive, got: 3
+
+    iex> Nx.Shape.validate_offset!({3, 4}, -4)
+    ** (ArgumentError) absolute value of offset must be less than breadth when negative, got: -4
+  """
+  def validate_offset!(shape, offset) do
+    {len, breadth} = shape
+
+    case offset do
+      i when i >= 0 and i < len ->
+        :ok
+
+      i when i >= 0 ->
+        raise ArgumentError, "offset must be less than length when positive, got: #{inspect(i)}"
+
+      i when i < 0 and -i < breadth ->
```

Comment:
Shouldn't the checks be using div(len, 2) and div(breadth, 2) instead?

From what I gathered, the main diagonal is offset 0, and you add 1 to the offset for each diagonal above the main and subtract 1 for the ones below, right?

If so, for a 3x4 matrix, for instance, you would have offsets -2, -1, 0 and 1

Reply:
> From what I gathered, the main diagonal is offset 0, and you add 1 to the offset for each diagonal above the main and subtract 1 for the ones below, right?

Right. In essence, you give a positive value to "move" the starting indices horizontally and a negative value to move it vertically.

So if I take your example of a 3 x 4 matrix:

```elixir
iex(2)> t = Nx.iota({3, 4})
#Nx.Tensor<
  s64[3][4]
  [
    [0, 1, 2, 3],
    [4, 5, 6, 7],
    [8, 9, 10, 11]
  ]
>
```

This matrix has 6 diagonals:

```elixir
iex(4)> Nx.take_diagonal(t, offset: -2)
#Nx.Tensor<
  s64[1]
  [8]
>
iex(5)> Nx.take_diagonal(t, offset: -1)
#Nx.Tensor<
  s64[2]
  [4, 9]
>
iex(6)> Nx.take_diagonal(t, offset: 0) 
#Nx.Tensor<
  s64[3]
  [0, 5, 10]
>
iex(7)> Nx.take_diagonal(t, offset: 1)
#Nx.Tensor<
  s64[3]
  [1, 6, 11]
>
iex(8)> Nx.take_diagonal(t, offset: 2)
#Nx.Tensor<
  s64[2]
  [2, 7]
>
```

So my current check makes sense.

Reply:
Yeah, the current implementation looks correct :)